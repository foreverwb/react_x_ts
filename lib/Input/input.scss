@import "../_main";

.ui-input {
  $border-radius: $input-border-radius;
  height: $input-height;
  border-radius: $border-radius;
  box-shadow: inset 0 1px 2px fade_out($black, 0.9);
  font-size: inherit;
  border: 1px solid $border-color; color: inherit;
  &:focus {
    outline: none;
    box-shadow: inset 0 1px 2px fade_out(darken($blue, 40%), 0.5);
  }
  &[disabled] {background: #ddd; color: #666; cursor: not-allowed; }
  &-wrapper {
    display: inline-flex; vertical-align: top; font-size: 14px;
    & + & {margin-left: 10px}
  }
}