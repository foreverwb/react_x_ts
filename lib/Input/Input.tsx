import React from 'react';
import {scopedClassMaker} from '../untils/classes';
import './input.scss'

const sc = scopedClassMaker('ui-input');

export interface IProps extends React.InputHTMLAttributes<HTMLInputElement> {
  length?: number | string;
  label?: string;
  labelPosition?: 'left' | 'top';
  error?: string;
  errorPosition?: 'right' | 'bottom'
}

const Input: React.FunctionComponent<IProps> = (props) => {
  const {length, type, label, labelPosition, error, errorPosition, ...restProps} = props;
  const width = length ? `calc(${length}em + 18px` : undefined;
  const c1 = labelPosition === 'top' ? 'block' : 'inline-block';
  const c2 = errorPosition === 'bottom' ? 'column' : '';
  return (
    <div className={sc('wrapper')}>
      {label && <label className={sc('label')} style={{display: c1}}>{label}</label>}
      <input
        type={type}
        {...restProps}
        className={sc('')}
        style={{width}}
      />
      {error && <span className={sc('error', {extra: c2})}>{error}</span>}
    </div>
  );
};
Input.defaultProps = {
  type: 'text'
};
export default Input;