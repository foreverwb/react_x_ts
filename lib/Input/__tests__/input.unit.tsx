import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../Input';

describe('Input', () => {
  it('prop label = hello', () => {
    const snap = renderer
      .create(<Input label="hello"/>)
      .toJSON();
    expect(snap).toMatchSnapshot();
  });
  it('prop labelPosition = top', () => {
    const snap = renderer
      .create(<Input label="hello" labelPosition="top"/>)
      .toJSON();
    expect(snap).toMatchSnapshot();
  });
  it('prop error = wrong', () => {
    const snap = renderer
      .create(<Input error="wrong"/>)
      .toJSON();
    expect(snap).toMatchSnapshot();
  });
});