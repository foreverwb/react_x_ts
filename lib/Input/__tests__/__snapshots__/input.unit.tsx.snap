// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input prop error = wrong 1`] = `
<div
  className="ui-input-wrapper"
>
  <input
    className="ui-input"
    style={
      Object {
        "width": undefined,
      }
    }
    type="text"
  />
  <span
    className="ui-input-error"
  >
    wrong
  </span>
</div>
`;

exports[`Input prop label = hello 1`] = `
<div
  className="ui-input-wrapper"
>
  <label
    className="ui-input-label"
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    hello
  </label>
  <input
    className="ui-input"
    style={
      Object {
        "width": undefined,
      }
    }
    type="text"
  />
</div>
`;

exports[`Input prop labelPosition = top 1`] = `
<div
  className="ui-input-wrapper"
>
  <label
    className="ui-input-label"
    style={
      Object {
        "display": "block",
      }
    }
  >
    hello
  </label>
  <input
    className="ui-input"
    style={
      Object {
        "width": undefined,
      }
    }
    type="text"
  />
</div>
`;
